tap_github/sync.py

```diff
@@ -95,6 +95,39 @@ def translate_state(state, catalog, repositories):
     nested_dict = lambda: collections.defaultdict(nested_dict)
     new_state = nested_dict()
 
+    # Collect keys(repo_name for update state or stream_name for older state) from state available in the `bookmarks``
+    previous_state_keys = state.get('bookmarks', {}).keys()
+    # Collect stream names from the catalog
+    stream_names = [stream['tap_stream_id'] for stream in catalog['streams']]
+
+    deselected_repos=True
+    # If the state contains bookmark for `repo_a` and `repo_b` and the user deselects these both repos and adds another repo
+    # then in that case this function was returning an empty state. Now this change will return the existing state instead of the empty state.
+   
+    for key in previous_state_keys:
+        # Loop through each key of `bookmarks` available in the previous state.
+
+        # Case 1:
+        # Check if it is the stream name or not. Older connections `bookmarks` contain stream names.
+        # Mark `deselected_repos` as `False` if at least one stream name is found in the previous state's keys because we want
+        # to migrate each stream's bookmark into the repo name.
+        # Example: {`bookmarks`: {`stream_a`: `bookmark_a`}} to {`bookmarks`: {`repo_a`: {`stream_a`: `bookmark_a`}}}
+
+        # Case 2:
+        # Check if the key is available in the list of currently selected repo's list or not. Newer format `bookmarks` contain repo names.
+        # Mark `deselected_repos` as `False` if at least one repo name matches the previous state's keys.
+
+        # At last, if `deselected_repos` remain True, then we will return the state itself. Because If the state contains a bookmark
+        # for `repo_a` and `repo_b` and the user deselects these both repos and adds another repo
+        # then in that case this function was returning an empty state. Now this change will return the existing state instead of the empty state.
+
+        if key in stream_names or key in repositories:
+            deselected_repos=False
+
+    # Return the existing state if all repos from the previous state are deselected(not found) in the current sync.
+    if deselected_repos:
```

Comment:
Can't we return on the first deselected repo? Because if we found any repo name, that itself indicates that the state file is in the new format.

Reply:
Updated logic.